### PR TITLE
ADD ignoreemptyservice option for kubernetes plugin

### PR DESCRIPTION
**Pull request by darkweaver87**
### 1. What does this pull request do?
add ignoreemptyservice option for kubernetes plugin. Example:

```
# host nginx        
nginx.dip-sre.svc.cluster1.local has address 10.32.241.140
```
now nginx doesn't have any pod remaining:
```
# host nginx        
nginx.dip-sre.svc.cluster2.local  has address 10.33.241.140
```
This way by only using search domains of resolv.conf, we are able to redirect a pod to another kubernetes cluster.

### 2. Which issues (if any) are related?
#1809 

### 3. Which documentation changes (if any) need to be made?
none

### CODE REVIEW

File: man/coredns-kubernetes.7 [hunk 1]
```diff
@@ -95,6 +95,9 @@ kubernetes [ZONES\.\.\.] {
 .IP "\(bu" 4
 \fBfallthrough\fR \fB[ZONES\.\.\.]\fR If a query for a record in the zones for which the plugin is authoritative results in NXDOMAIN, normally that is what the response will be\. However, if you specify this option, the query will instead be passed on down the plugin chain, which can include another plugin to handle the query\. If \fB[ZONES\.\.\.]\fR is omitted, then fallthrough happens for all zones for which the plugin is authoritative\. If specific zones are listed (for example \fBin\-addr\.arpa\fR and \fBip6\.arpa\fR), then only queries for those zones will be subject to fallthrough\.
 .
+.IP "\(bu" 4
+\fBignoreemptyservice\fR will turn off the serving of empty or failing services\. All empty services queries will result in an NXDOMAIN\.
```
Comment (chrisohaver) [inline man/coredns-kubernetes.7]: man pages are generated at release - so you don't need to modify them here.

File: plugin/kubernetes/README.md [hunk 2]
```diff
@@ -101,6 +101,8 @@ kubernetes [ZONES...] {
   the query. If **[ZONES...]** is omitted, then fallthrough happens for all zones for which the plugin
   is authoritative. If specific zones are listed (for example `in-addr.arpa` and `ip6.arpa`), then only
   queries for those zones will be subject to fallthrough.
+* `ignoreemptyservice` will turn off the serving of empty or failing services.
```
Comment (chrisohaver) [inline plugin/kubernetes/README.md]: "empty or failing" may be interpreted differently.  How about...

* `ignoreemptyservice` return `NXDOMAIN` for services without any ready endpoint addresses (e.g. ready pods).

**Review by chrisohaver — CHANGES_REQUESTED** [review]
Thanks for the PR. Looks good with a docs tweak.

Regarding the return code ... I'm thinking it may be more appropriate to return `NOERROR`, with no answers (aka "NODATA").  Per the [spec](https://github.com/kubernetes/dns/blob/master/docs/specification.md#241---a-records), "NODATA" is the correct response for a headless service with no ready endpoint addresses.  So, it seems we might want to follow suit there for cluster ip services.

**Comment by chrisohaver** [comment]
@miekg, are you OK with adding this new option to `kubernetes` plugin?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] ADD ignoreemp..." ]
>@miekg, are you OK with adding this new option to `kubernetes` plugin?

yep.

Should we rename it to: 'ignore emptyservice' ? To allow for future expansion?

**Comment by chrisohaver** [comment]
>  I'm thinking it may be more appropriate to return NOERROR, with no answers (aka "NODATA").

I re-read your description, and I realize now that the underlying need is based on resolv.conf search path.  An `NXDOMAIN` response is required to continue down the search path.

It seems that to be be consistent, you would also want to return `NXDOMAIN` for headless services without any ready endpoint addresses.

**Comment by chrisohaver** [comment]
> Should we rename it to: 'ignore emptyservice' ? To allow for future expansion?

Breaking it out certainly makes it more readable.

**Comment by chrisohaver** [comment]
Perhaps we should briefly document the use case too - since the feature is tailored to a specific use case. Would be good to hint at why you'd want to use the feature.  e.g. 

* `ignore emptyservice` return `NXDOMAIN` for services without any ready endpoint addresses (e.g. ready pods). This allows the querying pod to continue searching for the service in the search path.  The search path could, for example, include another kubernetes cluster.

**Comment by miekg** [comment]
@fastest963 seems corbot likes you :) I'll implement a different selection algo shortly that *first* looks in the most logical OWNERS file instead of piling them all up in randomly selecting someone

**Review by chrisohaver — CHANGES_REQUESTED** [review]
Looks good.  Please add unit tests (slipped my mind before).

**Comment by chrisohaver** [comment]
FYI - regarding headless service case, the kubernetes plugin currently replies with NXDOMAIN, i.e. it doesn't follow the spec as written. See #1814 - kube-dns also answers NXDOMAIN.

**Comment by youkoulayley** [comment]
Hi,

I have added the unit tests, but it failed on the integration tests. My local tests work, could you tell me why the tests fail and where I have made a mistake ? 

Thanks

**Comment by miekg** [comment]
Fails with:
~~~
== RUN   TestUpstreamToSelf
=== RUN   TestUpstreamToSelf/ext-svc.test-1.svc.cluster.local._A
--- FAIL: TestUpstreamToSelf (6.20s)
    --- FAIL: TestUpstreamToSelf/ext-svc.test-1.svc.cluster.local._A (0.28s)
    	helpers.go:123: answer for "ext-svc.test-1.svc.cluster.local." contained 1 results, 2 expected
    	helpers.go:301: ;; opcode: QUERY, status: NOERROR, id: 47731
    		;; flags: rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
    		
    		;; QUESTION SECTION:
    		;ext-svc.test-1.svc.cluster.local.	IN	 A
    		
    		;; ANSWER SECTION:
    		ext-svc.test-1.svc.cluster.local.	5	IN	CNAME	example.net.
    		
    	upstream_test.go:53: coredns log: 2018/05/22 09:33:49 [INFO] CoreDNS-1.1.2
    		2018/05/22 09:33:49 [INFO] linux/amd64, go1.10.2, 3db18b85
    		.:53
    		CoreDNS-1.1.2
    		linux/amd64, go1.10.2, 3db18b85
    		172.17.0.8:41112 - [22/May/2018:09:33:54 +0000] 23986 "A IN ext-svc.test-1.svc.cluster.local. udp 74 false 4096" NOERROR qr,aa,rd,ra 131 174.31µs
=== RUN   TestUpstreamToOther
=== RUN   TestUpstreamToOther/ext-svc.test-1.svc.cluster.local._A
--- PASS: TestUpstreamToOther (66.60s)
    --- PASS: TestUpstreamToOther/ext-svc.test-1.svc.cluster.local._A (60.74s)
=== RUN   TestUpstreamLoopBreak
=== RUN   TestUpstreamLoopBreak/upriver.test-2.svc.cluster.local._A
--- FAIL: TestUpstreamLoopBreak (6.24s)
    --- FAIL: TestUpstreamLoopBreak/upriver.test-2.svc.cluster.local._A (0.28s)
    	helpers.go:123: answer for "upriver.test-2.svc.cluster.local." contained 1 results, 2 expected
    	helpers.go:301: ;; opcode: QUERY, status: NOERROR, id: 60017
    		;; flags: rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
    		
    		;; QUESTION SECTION:
    		;upriver.test-2.svc.cluster.local.	IN	 A
    		
    		;; ANSWER SECTION:
    		upriver.test-2.svc.cluster.local.	5	IN	CNAME	up.river.local.
    		
    	upstream_test.go:139: coredns log: .:53
    		CoreDNS-1.1.2
    		linux/amd64, go1.10.2, 3db18b85
    		2018/05/22 09:35:02 [INFO] CoreDNS-1.1.2
    		2018/05/22 09:35:02 [INFO] linux/amd64, go1.10.2, 3db18b85
    		172.17.0.8:51674 - [22/May/2018:09:35:07 +0000] 62909 "A IN upriver.test-2.svc.cluster.local. udp 74 false 4096" NOERROR qr,aa,rd,ra 134 988.211µs
FAIL
FAIL	github.com/coredns/ci/test/kubernetes	216.993s
Makefile:74: recipe for target 'test-k8s' failed
make: *** [test-k8s] Error 1
~~~

**Comment by chrisohaver** [comment]
@darkweaver87, These don't appear to be related to your change.

I see the same tests failing consistently on another PR, which is also unrelated.
I'll check the recent commits, and for any recent activity on the integration test server.
The failed tests related to CNAME target resolution.

**Comment by chrisohaver** [comment]
@darkweaver87, thanks for the setup test.

If you can, please also add a test which tests the `ignore` functionality itself.

You can use `kubernetes/handler_test.go` as an example of how to do a test like this.
`kubernetes/handler_test.go` sets up mocked responses from the API, then runs a series of test queries through `ServeDNS()`.  In your case, you'd create a cluster ip service that has no endpoints for it, and turn the ignore feature on directly (e.g. `k.ignoreEmptyService = true`).

**Comment by youkoulayley** [comment]
Unit tests added, but the integration tests still failed on drone, is it normal ?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] ADD ignoreemp..." ]
>Unit tests added, but the integration tests still failed on drone, is it normal ?

I fix that in #1829.

**Comment by miekg** [comment]
little nit: can you, for constancy name put an underscore in emptyservice: `empty_service`

**Comment by youkoulayley** [comment]
> little nit: can you, for constancy name put an underscore in emptyservice: empty_service

Done

**Review by chrisohaver — APPROVED** [review]
looks great thanks!

**Comment by chrisohaver** [comment]
re-running /integration with  #1829 fix.

**Comment by chrisohaver** [comment]
> re-running /integration with #1829 fix.

nevermind - it hasnt merged yet.

**Comment by chrisohaver** [comment]
ok, now re-running /integration with #1829 fix merged.

**Comment by miekg** [comment]
Can someone put a small blurb on this feature in https://github.com/coredns/coredns/wiki/CoreDNS-1.1.3-Release-Notes ?

We want to release 1.1.3 today/tomorrow.

Thanks!

**Comment by chrisohaver** [comment]
> small blurb

done

**Pull request merged**